File: ARC104/B_DNA_Sequence/v1.scala
// AC

object Main extends App {
  val sc = new java.util.Scanner(System.in);
  val n = sc.nextInt();
  val s = sc.next();

  val answer = (1 to (n/2)).map(2 * _).map { m =>
    var k1 = (0 until m).count(i => s(i) == 'A');
    var k2 = (0 until m).count(i => s(i) == 'T');
    var k3 = (0 until m).count(i => s(i) == 'C');
    var k4 = (0 until m).count(i => s(i) == 'G');
    (0 to (n-m)).count { i =>
      if (i > 0) {
        s(i - 1) match {
          case 'A' => k1 -= 1;
          case 'T' => k2 -= 1;
          case 'C' => k3 -= 1;
          case 'G' => k4 -= 1;
          case _ => ;
        }
        s(i + m - 1) match {
          case 'A' => k1 += 1;
          case 'T' => k2 += 1;
          case 'C' => k3 += 1;
          case 'G' => k4 += 1;
          case _ => ;
        }
      }
      k1 == k2 && k3 == k4;
    }
  }.sum;

  println(answer);

}

File: abc171/d/v1.java
// AC

import java.util.Scanner;

class Main {
  public static void main(String[] args) {
    var sc = new Scanner(System.in);
    int n = sc.nextInt();
    var as = new int[n];
    for (int i = 0; i < n; i++) as[i] = sc.nextInt();
    int q = sc.nextInt();
    var bs = new int[q];
    var cs = new int[q];
    for (int i = 0; i < q; i++) {
      bs[i] = sc.nextInt();
      cs[i] = sc.nextInt();
    }

    var map = new java.util.HashMap<Integer, Integer>();
    long sum = 0;
    for (int i = 0; i < n; i++) {
      int a = as[i];
      sum += a;
      int count = map.getOrDefault(a, 0);
      map.put(a, count + 1);
    }
    for (int i = 0; i < q; i++) {
      int b = bs[i];
      int c = cs[i];
      int count = map.getOrDefault(b, 0);
      if (count > 0) {
        int count2 = map.getOrDefault(c, 0);
        map.put(b, 0);
        map.put(c, count + count2);
        sum += count * (c - b);
      }
      System.out.println(sum);
    }
  }
}
